cEP-0025.md

```diff
@@ -0,0 +1,116 @@
+
+## Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. There will be a separate repository named as coala-pyflakes which will
+   be installable via `pip install coala-pyflakes`.
+2. The repository will contain a new package `pyflakes-plugins` which
+   would house all bears which uses pyflakes-enhanced-ast.
+3. The `pyflakes-plugins` would contain a metabear `PyflakesASTBear`
+   which would be used by all other plugin bears.
+
+## Management of the new `coala-pyflakes` repository
+
+The repository will use a similar mechanism of what is being
+used in `coala-bears` to test all its plugins.
+
+## Code Samples/Prototypes
+
+Here is a prototype for the implementations of PyflakesASTBear:
+
+```python
+# Imports not mentioned to maintain brevity
+
+class PyflakesResult(HiddenResult):
+
+    @enforce_signature
+    def __init__(self, origin, deadScopes):
+
+        Result.__init__(self, origin, message='')
+
+        self.class_scope = self.get_scope(ClassScope, deadScopes)
+        self.function_scope = self.get_scope(FunctionScope, deadScopes)
+        self.module_scope = self.get_scope(ModuleScope, deadScopes)
+        self.generator_scope = self.get_scope(GeneratorScope, deadScopes)
+        self.doctest_scope = self.get_scope(DoctestScope, deadScopes)
+
+    def get_scope(self, scope_level, scopes):
+        return list(filter(lambda scope: isinstance(scope, scope_level),
+                           scopes))
+
+    def get_node_location(self, node):
+        return (node.source.lineno,
+                node.source.lineno + node.source.depth - 1,
+                node.source.col_offset)
+
+
+class PyflakesASTBear(LocalBear):
+
+    def run(self, filename, file):
+        """
+        Generates pyflakes-enhance-AST for the input file and returns
+        deadScopes as HiddenResult
+        :return:
+            One HiddenResult containing a dictionary with keys being
+            type of scope and values being a list of scopes generated
+            from the file.
+        """
+        tree = ast.parse(''.join(file))
+        result = Checker(tree, filename)
+
+        yield PyflakesResult(self, result.deadScopes)
+```
+
+Here is a prototype for a plugin that uses PyflakesASTBear:
+
+```python
+# Imports not mentioned to maintain brevity
+
+class FutureImportBear(LocalBear):
+    """
+    Uses pyflakes-enhance-AST to get list of future imports
+    """
+
+    BEAR_DEPS = {PyflakesASTBear}
+
+    def run(self, filename, file,
+            dependency_results=dict()
+            ):
+        for result in dependency_results.get(PyflakesASTBear.name, []):
+            for scope in result.module_scope:
+                for node_name, node in scope.items():
+                    if isinstance(node, FutureImportation):
+                        yield Result.from_values(
+                            origin=self,
+                            message=('Future import %s found' % node_name),
```

Comment:
this example assumes creates an error for any future imports.

pls investigate the new py37 future import, to see if it is desirable.

If yes, the logic needs to be improved.

If not, then this is a `NoFutureImportBear`.

Reply:
I was planning to make this bear work 2 ways: 
1. To make sure that a particular import is added to every module. For eg: If I want that print_function should be imported into every module, I would make it compulsory and use this to find instances where I have failed to add them.
2. If I don't want them, or don't want some (create a blacklist). The NoFutureImportBear you are discussing.

Reply:
As an example `NoFutureImportBear` is better.
It is a simple bear that coala can use, as there are no py3 future imports which are supported on py34.  Just add a comment which explains why this bear is simple.
An objective of this cEP and project is to show that it will allow simple rules based on the pyflakes ast.
Making it a complicated bear detracts from that goal.